Adding support for /proc/buddyinfo for linux free memory fragmentation.

thorhs opened the pull request:
/prod/buddyinfo returns data on the free blocks fragments available
for use from the kernel.  This data is useful when diagnosing
possible memory fragmentation.

More info can be found in:
* https://lwn.net/Articles/7868/
* https://andorian.blogspot.com/2014/03/making-sense-of-procbuddyinfo.html

grobie commented on buddyinfo.go:
```diff
@@ -0,0 +1,99 @@
+// limitations under the License.
+
+// +build !nobuddyinfo
+// +build !windows,!netbsd
```
Procfs is a library and doesn't use build tags so far. Let's remove these two lines.

grobie commented on buddyinfo.go:
```diff
@@ -0,0 +1,99 @@
+	"strings"
+)
+
+// A BuddyInfo is the details parsed from /proc/[pid]/buddyinfo.
```
This seems to be wrong, buddyinfo is not process specific, is it?

thorhs replied:
Copy/Paste strikes again.  Fixed, thanks.

grobie commented on buddyinfo.go:
```diff
@@ -0,0 +1,99 @@
+// It is a accessed by node and zone.
+// The data is comprised of an array of free fragments of each size.
+// The sizes are 2^n*PAGE_SIZE, where n is the array index.
+type BuddyInfo map[string]map[string][]float64
```
Is a map really the best data structure here?

thorhs replied:
When I started, I thought I might need to do more processing where it would be good to directly address each line.  Thinking about it, this could probably be an array of structs, which contain:
* node
* zone
* array of fragment counts

Should I change to that implementation, it should be relatively simple to implement.

grobie commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,66 @@
+	}
+	defer file.Close()
+
+	buddyInfo, err := parseBuddyInfo(file)
```
Let's not use internal package functions in tests, this would have catched your last fix automatically. Instead use the `fs.NewBuddyInfo()` function directly and put the fixtures in different folders, e.g. `FS("fixtures/buddyinfo/short").NewBuddyInfo()`.

thorhs commented:
I hope I have addressed all the concerns raised, except the data structure.

For the external facing data structure, I'm considering two options. Both use BuddyInfoRec as the storage of data:
```golang
  typedef BuddyInfo struct {
    node string
    zone string
    sizes []float64
  }
```

The question becomes, should I wrap it into a struct, or a simple typedef:

```golang
  typedef BuddyInfo []BuddyInfoRec
```
or:

```golang
  typedef BuddyInfo struct {
    Records []BuddyInfoRec
  }
```

mdlayher commented:
Out of curiosity, are the "fields" always named "DMA", "DMA32", and "Normal"?

mdlayher commented on buddyinfo.go:
```diff
@@ -0,0 +1,96 @@
+		buddyInfo[node][zone] = sizes
+	}
+
+	return buddyInfo, nil
```
You'll want to `return buddyInfo, scanner.Err()` in case the loop breaks early and an error is reported.

mdlayher commented on buddyinfo.go:
```diff
@@ -0,0 +1,96 @@
+			bucketCount = arraySize
+		} else {
+			if bucketCount != arraySize {
+				return nil, fmt.Errorf("Mismatch in number of buddyinfo buckets, previous count %d, new count %d", bucketCount, arraySize)
```
Error should start with a lowercase "m".

mdlayher commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,47 @@
+func TestBuddyInfoShort(t *testing.T) {
+	_, err := FS("fixtures/buddyinfo/short").NewBuddyInfo()
+	if err == nil {
+		t.Errorf("expected error, but non occurred")
```
*none

mdlayher commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,47 @@
+func TestBuddyInfoSizeMismatch(t *testing.T) {
+	_, err := FS("fixtures/buddyinfo/sizemismatch").NewBuddyInfo()
+	if err == nil {
+		t.Errorf("expected error, but non occurred")
```
*none

grobie commented:
On Tue, Feb 7, 2017, 11:02 Þórhallur Sverrisson <notifications@github.com>
wrote:

> I hope I have addressed all the concerns raised, except the data structure.
>
> For the external facing data structure, I'm considering two options. Both
> use BuddyInfoRec as the storage of data:
>
>   typedef BuddyInfo struct {
>     node string
>     zone string
>     sizes []float64
>   }
>
> The question becomes, should I wrap it into a struct, or a simple typedef:
>
What about keeping it simple and using `([]BuddyInfo, error)`?

>   typedef BuddyInfo []BuddyInfoRec
>
> or:
>
>   typedef BuddyInfo struct {
>     Records []BuddyInfoRec
>   }
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/procfs/pull/34#issuecomment-278044493>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaADB4kxLM4C1siTnxm_PgtgG14NDks5raJU0gaJpZM4L5hjY>
> .
>

thorhs commented:
Hope this addresses the concerns you had, it is much cleaner now.

Sorry for the churn.

grobie commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func TestBuddyInfoShort(t *testing.T) {
+	_, err := FS("fixtures/buddyinfo/short").NewBuddyInfo()
```
I don't think this is testing the desired error condition, as the fixture file isn't named `buddyinfo`. It seems to make sense to verify that the returned error message includes the expected message.

thorhs replied:
Fixed and tested.

grobie commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func TestBuddyInfoSizeMismatch(t *testing.T) {
+	_, err := FS("fixtures/buddyinfo/sizemismatch").NewBuddyInfo()
```
See above.

grobie commented on buddyinfo.go:
```diff
@@ -0,0 +1,96 @@
+		scanner   = bufio.NewScanner(r)
+	)
+
+	bucketCount := -1
```
I wouldn't use two different ways to initialize variables right next to each other and move this line into the var block above.

mdlayher commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	if want, got := "invalid number of fields when parsing buddyinfo", err.Error(); want != got {
+		t.Errorf("wrong error returned, wanted '%s', got '%s'", want, got)
```
Use `%q` instead of `'%s'` as it'll handle weird corner cases for quoting and escaping for you.

mdlayher commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	if want, got := "mismatch in number of buddyinfo buckets", err.Error(); !strings.HasPrefix(got, want) {
+		t.Errorf("wrong error returned, wanted prefix '%s', got '%s'", want, got)
```
Use `%q` instead of `'%s'` as it'll handle weird corner cases for quoting and escaping for you.

mdlayher commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	if want, got := "invalid number of fields when parsing buddyinfo", err.Error(); want != got {
+		t.Errorf("wrong error returned, wanted '%q', got '%q'", want, got)
```
Can also remove the single quotes around `%q`.

mdlayher commented on buddyinfo_test.go:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	if want, got := "mismatch in number of buddyinfo buckets", err.Error(); !strings.HasPrefix(got, want) {
+		t.Errorf("wrong error returned, wanted prefix '%q', got '%q'", want, got)
```
Can also remove the single quotes around `%q`.

thorhs replied:
Ohh, live and learn :)  Never used %q before.

grobie commented:
Thanks!
